File: Algorithm/数组/查找/BinarySearch.py
class Solution:
    def binarySearch(self, nums: list[int], target: int) -> int:
        left, right = 0, len(nums) - 1
        while left <= right:
            middle = left + (right - left) // 2
            if nums[middle] > target:
                right = middle - 1
            elif nums[middle] < target:
                left = middle + 1
            else:
                return middle
        return -1

    def binarySearch2(self, nums: list[int], target: int) -> int:
        left, right = 0, len(nums)
        while left < right:
            middle = left + (right - left) // 2
            if nums[middle] > target:
                right = middle
            elif nums[middle] < target:
                left = middle + 1
            else:
                return middle
        return -1

if __name__=="__main__":
    nums = [-1,0,3,5,9,12]
    target = 3
    print(Solution().binarySearch(nums, target))
    print(Solution().binarySearch2(nums, target))

File: Algorithm/leetcode/practice/数组/查找/BinarySearch.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution
{
public:
	int binary_search1(vector<int> &nums, int target)
	{
		int left = 0;
		int right = nums.size() - 1;
		while (left <= right)
		{
			int middle = left + (right - left) / 2;
			if (nums[middle] == target)
			{
				return middle;
			}
			else if (nums[middle] < target)
			{
				left = middle + 1;
			}
			else if (nums[middle] > target)
			{
				right = middle - 1;
			}
		}
		return -1;
	};
	int binary_search2(vector<int> &nums, int target)
	{
		int left = 0;
		int right = nums.size();
		while (left < right)
		{
			int middle = left + (right - left) / 2;
			if (nums[middle] == target)
			{
				return middle;
			}
			else if (nums[middle] < target)
			{
				left = middle + 1;
			}
			else if (nums[middle] > target)
			{
				right = middle;
			}
		}
		return -1;
	};
};

int main()
{
	vector<int> nums = {1, 2, 3, 4, 5, 6};
	int target = 3;
	Solution s;
	int res = s.binary_search2(nums, target);
	cout << res << endl;
	return 0;
}